### PR TITLE
Updating endpoints only does not require previous snapshot

### PR DESCRIPTION
**Pull request by npolshakova**
# Description

If there are user config errors that would normally prevent Gloo from starting up, we still need the control plane to be able to serve EDS updates even when the snapshot cache has been cleared. EDS updates should occur in all scenarios when the pod is running, regardless if there is a previous snapshot.

# Context

Need to ensure Gloo pod starts and serves EDS regardless of config errors when the pod is bounced and snapshot cache is cleared. EDS updates should occur in all scenarios when the pod is running, even when the snapshot cache has been cleared (bouncing gloo pod, settings change, etc.) 

https://github.com/solo-io/gloo/issues/4345

# Checklist:

- [x] I included a concise, user-facing changelog (for details, see https://github.com/solo-io/go-utils/tree/master/changelogutils) which references the issue that is resolved.
- [x] If I updated APIs (our protos) or helm values, I ran `make install-go-tools generated-code` to ensure there will be no code diff
- [x] I followed guidelines laid out in the Gloo Edge [contribution guide](https://docs.solo.io/gloo-edge/latest/contributing/)
- [x] I opened a draft PR or added the work in progress label if my PR is not ready for review
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
BOT NOTES: 
resolves https://github.com/solo-io/gloo/issues/4345

### CODE REVIEW

File: test/kube2e/gateway/robustness_test.go [hunk 1]
```diff
@@ -108,16 +133,32 @@ var _ = Describe("Robustness tests", func() {
...
+				return len(services.Items) == 0
+			}, "15s", "0.5s").Should(BeTrue())
+		}
 
 		cancel()
 	})
 
 	It("updates Envoy endpoints even if proxy is rejected", func() {
 
-		By("create a deployment and a matching service")
```
Comment (yuval-k) [inline test/kube2e/gateway/robustness_test.go:117]: why was this removed?

File: projects/gloo/pkg/syncer/translator_syncer_test.go [hunk 2]
```diff
@@ -337,8 +459,14 @@ func (t *mockTranslator) Translate(params plugins.Params, proxy *v1.Proxy) (envo
 				}
 			}
 		}
+		if t.currentSnapshot != nil {
```
Comment (yuval-k) [inline projects/gloo/pkg/syncer/translator_syncer_test.go:462]: in the code this part is in the syncer. why is it in mock translator here? should it be in level above?

File: test/kube2e/gateway/robustness_test.go (hunk 1 above)
Comment (npolshakova) [reply test/kube2e/gateway/robustness_test.go:117]: I moved it to the BeforeEach. Now each test sets up the appDeployment and appService before the test, and then deletes it after.

File: projects/gloo/pkg/syncer/translator_syncer_test.go (hunk 2 above)
Comment (npolshakova) [reply projects/gloo/pkg/syncer/translator_syncer_test.go:462]: This is just the test to be able to check the non NilSnapshot case. Otherwise syncEnvoy will use the NilSnapshot instead when it tests Sync (https://github.com/solo-io/gloo/blob/master/projects/gloo/pkg/syncer/envoy_translator_syncer.go#L125). When the mock translator has a snapshot set it will use that and we can verify only the endpoints/clusters were updated.

File: projects/gloo/pkg/xds/envoy_snapshot.go [hunk 3]
```diff
@@ -93,6 +93,19 @@ func NewSnapshotFromResources(
 	}
 }
 
+func NewEndpointsSnapshotFromResources(
+	endpoints cache.Resources,
+	clusters cache.Resources,
+) cache.Snapshot {
+	// TODO: Copy resources and downgrade, maybe maintain hash to not do it too many times
```
Comment (mlholland) [inline projects/gloo/pkg/xds/envoy_snapshot.go:100]: Is there an issue link for this? If so, can you include the link here?

File: test/kube2e/gateway/robustness_test.go [hunk 4]
```diff
@@ -254,6 +295,172 @@ var _ = Describe("Robustness tests", func() {
...
+			UpstreamType: &gloov1.Upstream_Static{
+				Static: &static_plugin_gloo.UpstreamSpec{
+					Hosts: []*static_plugin_gloo.Host{{
+						Addr: "localhost",
+						Port: 1234,
+					}},
+				},
+			},
+		}
+		_, err := upstreamClient.Write(upstream, clients.WriteOpts{Ctx: ctx, OverwriteExisting: true})
```
Comment (mlholland) [inline test/kube2e/gateway/robustness_test.go:315]: Looks like this err is unchecked.

File: projects/gloo/pkg/xds/envoy_snapshot.go (hunk 3 above)
Comment (npolshakova) [reply projects/gloo/pkg/xds/envoy_snapshot.go:100]: There isn't right now but I can make one. I think I'll remove this comment and just keep the one in NewSnapshotFromResources so we don't have it in two spots.

File: test/kube2e/gateway/robustness_test.go (hunk 4 above)
Comment (npolshakova) [reply test/kube2e/gateway/robustness_test.go:315]: Nice catch!

**Review by mlholland — APPROVED** [review]
